File: 数学/287.寻找重复数.py

# 暴力 n(log(n) + 1) 擦边过的
class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        nums.sort()
        hold = nums[0]
        for i in range(1, len(nums)):
            if hold == nums[i]:
                return hold
            hold = nums[i]

File: 数学/287.寻找重复数.js
/**
 * @param {number[]} nums
 * @return {number}
 * Floyd 判圈法（快慢指针），不修改原数组，O(n) 时间 O(1) 空间
 */
var findDuplicate = function (nums) {
  let slow = nums[0]
  let fast = nums[0];
  // 第一阶段：快慢指针相遇
  do {
    slow = nums[slow];
    fast = nums[nums[fast]];
  } while (slow !== fast);
  // 第二阶段：找到入口点
  slow = nums[0];
  while (slow !== fast) {
    slow = nums[slow];
    fast = nums[fast];
  }
  return slow;
};
